Inclinometer Plugin, Incremental Velocity Message, Edits to CMake/xac…

colefisher1 opened the pull request:
I wrote a Gazebo plugin to simulate the inclinometers on the new STIM300 IMU. I was unable to find a standard deviation value for the inclinometers' expected gaussian noise, so I simulated noise using the same standard deviation value (0.08825985) and noise streams (IMU_LINACC_NOISE...) as the IMU's accelerometers.

The new message I created stores a Header, Vector3 for XYZ incremental velocity, and an array for the covariance matrix.

MarshallRawson commented on mil_common/simulation/mil_gazebo/src/mil_inclinometer_gazebo.cpp:
```diff
@@ -0,0 +1,69 @@
+  update_connection_ = sensor_->ConnectUpdated(boost::bind(&MilInclinometerGazebo::OnUpdate, this));
+}
+
+void MilInclinometerGazebo::OnUpdate()
```
Make sure that this will only report incremental velocity at time intervals possible by the physical IMU, and make that amount configurable in the xacro (so that if we switch to another IMU / reconfigure this one to run at a diffrent clock speed, the data still makes sense).

colefisher1 replied:
I've changed it so the plugin reads in update_rate from the xacro and plugs it in to ros::rate loop_rate(update_rate). I've noticed that other plugins do not do this, is this the proper way of doing this?

MarshallRawson replied:
ya so typically, gazebo plugin use gazebo time not ros time. Here is an example of another plugin doing this. https://github.com/uf-mil/vrx/blob/8c204f359effb8ab810fac6b57ebc46d13dcaf3a/vrx_gazebo/src/light_buoy_plugin.cc#L227

colefisher1 commented on mil_common/simulation/mil_gazebo/src/mil_inclinometer_gazebo.cpp:
```diff
@@ -0,0 +1,69 @@
+  msg_.incremental_velocity_covariance[8] =
+      NoiseCovariance(*sensor_->Noise(gazebo::sensors::SensorNoiseType::IMU_LINACC_Z_NOISE_METERS_PER_S_SQR));
+
+  pub_ = nh_.advertise<mil_msgs::IncrementalVelocity>("/imu/inclinometer", 20);
```
Which particular element in the code you highlighted should I pass in via xacro?

MarshallRawson replied:
The topic name and the queue size

colefisher1 commented:
Updates: 
**mil_utils_gazebo.cpp**
-  Added an addRandomWalkNoise function to simulate noise.

**mil_inclinometer_gazebo.cpp**
- Made topic name, queue size, velocity random walk, and correlation time imported from xacro
- Update_rate is automatically applied to the plugin by the Sensor Class (no need to manually apply it)
- Called addRandomWalkNoise from gazebo_utils to properly add noise to the raw data

**IncrementalLinearVelocity.msg**
- Renamed to specify IncrementalLinearVelocity
- Added a non-stamped version of the message according to ROS standards
